### PR TITLE
Add PCE server protocol parameters

### PR DESCRIPTION
**Pull request by crcampos612**
Add PCE Protocol parameters at global level for PCE servers

Addition of the `PCEP` as a protocol in the Network Instance configuration

Included the authentication as an option using the latest `openconfig-key-chain` definition. 

```
module: openconfig-network-instance
  +--rw network-instances
     +--rw network-instance* [name]
        +--rw name                       -> ../config/name
        +--rw protocols
           +--rw protocol* [identifier name]
              +--rw identifier          -> ../config/identifier
              +--rw name                -> ../config/name
              +--rw pcep
                 +--rw path-computation-servers
                    +--rw path-computation-server* [pce-server-address]
                       +--rw pce-server-address    -> ../config/pce-server-address
                       +--rw config
                       |  +--rw pce-server-address?         oc-inet:ip-address
                       |  +--rw id?                         string
                       |  +--rw port?                       oc-inet:port-number
                       |  +--rw source-address?             oc-inet:ip-address
                       |  +--rw pce-type?                   pce-mode-type
                       |  +--rw preference?                 uint8
                       |  +--rw report-local-lsp?           boolean
                       |  +--rw pce-initiated-capability?   boolean
                       |  +--rw sr-support?                 boolean
                       +--ro state
                       |  +--ro pce-server-address?         oc-inet:ip-address
                       |  +--ro id?                         string
                       |  +--ro port?                       oc-inet:port-number
                       |  +--ro source-address?             oc-inet:ip-address
                       |  +--ro pce-type?                   pce-mode-type
                       |  +--ro preference?                 uint8
                       |  +--ro report-local-lsp?           boolean
                       |  +--ro pce-initiated-capability?   boolean
                       |  +--ro sr-support?                 boolean
                       +--rw timers
                       |  +--rw config
                       |  |  +--rw keepalive?                       uint8
                       |  |  +--rw dead-timer?                      uint8
                       |  |  +--rw redelegation-timeout-interval?   uint16
                       |  |  +--rw state-timeout-interval?          uint16
                       |  +--ro state
                       |     +--ro keepalive?                       uint8
                       |     +--ro dead-timer?                      uint8
                       |     +--ro redelegation-timeout-interval?   uint16
                       |     +--ro state-timeout-interval?          uint16
                       +--rw authentication
                          +--rw config
                          |  +--rw enable?               boolean
                          |  +--rw authentication-key?   oc-types:routing-password
                          |  +--rw keychain?             oc-keychain:keychain-ref
                          +--ro state
                             +--ro enable?               boolean
                             +--ro authentication-key?   oc-types:routing-password
                             +--ro keychain?             oc-keychain:keychain-ref
````

For the configuration of the protocol, we have followed the attributes supported by the following vendors in their configuration guides:

- Cisco PCEP configuration [Link](https://rendoaw.github.io/2016/12/Cisco-IOS-XR-Config-Example-for-PCEP-and-BGP-LS)
- Juniper PCEP configuration [Link](https://www.juniper.net/documentation/us/en/software/junos/mpls/topics/topic-map/pcep-configuration.html#id-example-configuring-the-path-computation-element-protocol-for-mpls-rsvp-te)

### CODE REVIEW

**Comment by s19nal** [comment]
hey @oscargdd this looks good to me. apologies for the very short review comment! :)

**Comment by s19nal** [comment]
@sbarguil @oscargdd   the only thing right now I'd want to add right now is authentication options?  Supported by multiple vendors, and I think we have some existing patterns for authentication that could be used here too? 
other than that it LGTM

**Comment by crcampos612** [comment]
> @sbarguil @oscargdd the only thing right now I'd want to add right now is authentication options? Supported by multiple vendors, and I think we have some existing patterns for authentication that could be used here too? other than that it LGTM

Added authentication container for enabling it, either through a MD5 password(key) or a reference to a keychain

**Comment by dplore** [comment]
@chrisy will review

**Review by chrisy — CHANGES_REQUESTED** [review]
Remove the `.DS_Store` file in `release/models`. I'll review the rest of the content shortly.

File: release/models/pcep/openconfig-pcep.yang
Comment (chrisy) [inline release/models/pcep/openconfig-pcep.yang:26]:
```suggestion
    relating to Path Computation Element Protocol (PCEP) for communications
```

File: release/models/pcep/openconfig-pcep.yang
Comment (chrisy) [inline release/models/pcep/openconfig-pcep.yang:195]: Could you fix the indentation on this and the subsequent descriptions, the quote mark should be 2 spaces deeper than `description`.

File: release/models/pcep/openconfig-pcep.yang
Comment (chrisy) [inline release/models/pcep/openconfig-pcep.yang:288]: More indentation to fix here and next description

File: release/models/pcep/openconfig-pcep.yang
Comment (chrisy) [inline release/models/pcep/openconfig-pcep.yang:205]: Could you check whether we more consistently call this `source-address` in other places? I don't think we refer to the originating address as `client-address` anywhere else, so I would prefer `source-address` I also don't think it makes sense to prepend with `pce-` when it's the PCC's address. Just `source-address` I think is a better fit.

**Review by chrisy — CHANGES_REQUESTED** [review]
Some minor indentation cleanup and similar, but also some quibbling over how to refer to `source-address`.

File: release/models/pcep/openconfig-pcep.yang
Comment (sbarguil) [reply release/models/pcep/openconfig-pcep.yang]: Sure! Done

File: release/models/pcep/openconfig-pcep.yang
Comment (sbarguil) [reply release/models/pcep/openconfig-pcep.yang]: Done

File: release/models/pcep/openconfig-pcep.yang
Comment (chrisy) [reply release/models/pcep/openconfig-pcep.yang:205]: > Could you check whether we more consistently call this `source-address` in other places? I don't think we refer to the originating address as `client-address` anywhere else, so I would prefer `source-address` I also don't think it makes sense to prepend with `pce-` when it's the PCC's address. Just `source-address` I think is a better fit.

This is the only comment I have outstanding. I did a quick check, we near universally refer to this as `source-address` everywhere else in the models; unless there's a strong reason to keep the proposed name, I'd like to see this be `source-address` also.

- 18 references to `source-address`
- 1 to `local-source-address` (in telemetry).
- 2 to `source-interface`

Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang]: Changed to source-address as proposed

**Review by chrisy — APPROVED** [review]
LGTM

**Comment by chrisy** [comment]
@crcampos612 Looks like you have merge conflicts to resolve, and my permissions did not transit from the models repo so we still will need someone else to approve this before it can be merged.

File: release/models/network-instance/openconfig-network-instance.yang
Comment (robshakir) [inline release/models/network-instance/openconfig-network-instance.yang:867]: Additional blank line?

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:116]:
```suggestion
          "PCC (Router) delegates control over LSP path to PCE (stateful active)";
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:120]:
```suggestion
          "PCC (Router) don't allow or revoke control over LSP path to
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:125]:
```suggestion
          "PCC (Router) only inform to a PCE whenever the state of an LSP changes,
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:142]:
```suggestion
        "Defines PCE server parameters.";
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:148]:
```suggestion
          "Configuration and state information for communication with a PCE server.";
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:138]:
```suggestion
      "Configuration and state data for establishing a PCEP session with PCE servers.";
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:155]:
```suggestion
            "IPv4 address of PCE server.";
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:187]: Suggest something like "specify multiple entries in the path-computation-server list to provide redundancy." rather than "repeat this step".

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:192]: Could this just be `id`, since the list is already called `path-computation-server`?

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:195]:
```suggestion
        "A unique name for the PCE server.";
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:198]: Again, could this be called `port`?

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:214]:
```suggestion
	    "The type of PCEP capability supported which is advertised in the Open
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:236]: I think we've generally called these `enable-...` rather than `x-enable`. Would appreciate other's thoughts here too.

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:240]:
```suggestion
      	 "Indicates to PCE that PCC (Router) supports PCE-initiated LSP paths instantiation.
```

File: release/models/pcep/openconfig-pcep.yang
Comment (robshakir) [inline release/models/pcep/openconfig-pcep.yang:351]:
```suggestion
        "Password (key) used for securing a PCEP session using
```

**Review by robshakir — CHANGES_REQUESTED** [review]
Thanks for all the work here, this generally LGTM - but I have a few comments/suggestions.

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang:294]: As per RFC 5440 for keepalive timer "The recommended defaut value is 30 seconds" and for deadtimer "is 4 times the value of the keepalive timer"

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang:347]: Done. See 2838b0e7bfe2cec7e430858e5d9abbbdc62aac3f

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang:357]: Done. See 2838b0e7bfe2cec7e430858e5d9abbbdc62aac3f

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang]: Done. See 2838b0e7bfe2cec7e430858e5d9abbbdc62aac3f

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang:347]: Done. See 2838b0e7bfe2cec7e430858e5d9abbbdc62aac3f

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang:357]: Done. See 2838b0e7bfe2cec7e430858e5d9abbbdc62aac3f

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang:294]: As per RFC 544, keepalive timer "the recommended default value is 30 seconds"

File: release/models/mpls/openconfig-mpls-static.yang
Comment (crcampos612) [reply release/models/mpls/openconfig-mpls-static.yang]: Changed to version 3.3.0. See b18facd9ad227af7d4bfac8cf32a4ac1f49934aa

File: release/models/network-instance/openconfig-network-instance.yang
Comment (crcampos612) [reply release/models/network-instance/openconfig-network-instance.yang]: Blank line removed. See 29e061f5157dbd5c4ca6da39894d43982742b9ae

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang]: Changed

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang]: Changed

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang]: Modified to "pce-initiated-capability" only, because the name autodescribe functionality by itself with two possible states, true or false, with the default option of "false". See falseccd5ed7102c5ab780f2ac12b2a2f76bb2ac92220

File: release/models/pcep/openconfig-pcep.yang
Comment (crcampos612) [reply release/models/pcep/openconfig-pcep.yang]: Actually, more than capabilities, there are three functionalities or features to be enable/disable in a router to talk PCEP with a PCE server:
- Report to PCE local LSPs
- Allow PCE-Initiated tunnels instantiation
- Support for Segment Routing tunnels
Because name autodescribe functionality by itself with two possible states, true or false, with the default option of "false", maybe it's not worth adding a specific container only for this. But, if so I change it.
See ccd5ed7102c5ab780f2ac12b2a2f76bb2ac92220

**Comment by dplore** [comment]
@sbarguil thank you for this contribution!  Is there reference vendor documentation you can point us to for the types and leafs added?

If you don't mind, can you also provide tree output showing the change?

**Comment by dplore** [comment]
I will dismiss rob's request for changes as I confirmed they have been met.

**Comment by sbarguil** [comment]
@dplore is there any pending point in this PR?

**Comment by sbarguil** [comment]
@dplore is there any pending point in this PR?

**Comment by dplore** [comment]
Thank you for the contribution @sbarguil  and thanks to all who contributed and @chrisy for the review.  

Author: Carlos Rodriguez Campos <carlos.rodriguezcampos.ext@telefonica.com>
Date:   Wed Jul 13 03:03:19 2022 +0000

    Add PCE server protocol parameters (#581)
    
    Add PCE Protocol parametrization at global level for servers and LSP control
    
    * Add PCE Protocol parameters at global level for PCE servers
    * Add PCEP as a protocol in the Network Instance configuration
    * Add authentication keychain for PCEP
    
    * (M) release/models/pcep/openconfig-pcep.yang
    * (M) release/models/mpls/openconfig-mpls-te.yang
    * (M) openconfig-mpls-static.yang
    * (M) openconfig-mpls-te.yang
    * (M) openconfig-mpls.yang
    * (M) openconfig-network-instance.yang
    * (M) openconfig-network-instance-l2.yang
    
    Co-authored-by: Óscar González de Dios <oscar.gonzalezdedios@telefonica.com>
    Co-authored-by: Chris Luke <chrisy@flirble.org>
    Co-authored-by: Rob Shakir <robjs@google.com>

**Pull request merged**
